midi/drivers/coremidi.c

```diff
@@ -0,0 +1,415 @@
+/*  RetroArch - A frontend for libretro.
+ *  Copyright (C) 2025 The RetroArch team
+ *
+ *  RetroArch is free software: you can redistribute it and/or modify it under the terms
+ *  of the GNU General Public License as published by the Free Software Found-
+ *  ation, either version 3 of the License, or (at your option) any later version.
+ *
+ *  RetroArch is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *  PURPOSE.  See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with RetroArch.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <CoreMIDI/CoreMIDI.h>
+#include <libretro.h>
+#include <lists/string_list.h>
+#include <string/stdstring.h>
+
+#include "../midi_driver.h"
+#include "../../verbosity.h"
+
+#define CORE_MIDI_QUEUE_SIZE 1024
+
+typedef struct {
+    midi_event_t events[CORE_MIDI_QUEUE_SIZE]; ///< Event buffer
+    int read_index;         ///< Current read position
+    int write_index;        ///< Current write position
+} coremidi_queue_t;
+
+typedef struct {
+    MIDIClientRef client;       ///< CoreMIDI client
+    MIDIPortRef input_port;     ///< Input port for receiving MIDI data
+    MIDIPortRef output_port;    ///< Output port for sending MIDI data
+    MIDIEndpointRef input_endpoint; ///< Selected input endpoint
+    MIDIEndpointRef output_endpoint; ///< Selected output endpoint
+    coremidi_queue_t input_queue; ///< Queue for incoming MIDI events
+} coremidi_t;
+
+/// Write to the queue
+static bool coremidi_queue_write(coremidi_queue_t *q, const midi_event_t *ev) {
+    int next_write = (q->write_index + 1) % CORE_MIDI_QUEUE_SIZE;
+    if (next_write == q->read_index) // Queue full
+        return false;
+
+    memcpy(&q->events[q->write_index], ev, sizeof(*ev));
+    q->write_index = next_write;
+    return true;
+}
+
+/// MIDIReadProc callback function
+static void midi_read_callback(const MIDIPacketList *pktlist, void *readProcRefCon, void *srcConnRefCon) {
+    coremidi_t *d = (coremidi_t *)readProcRefCon;
+    const MIDIPacket *packet = &pktlist->packet[0];
+    midi_event_t event;
+    uint32_t i;
+
+    for (i = 0; i < pktlist->numPackets; i++) {
+        const uint8_t *data = packet->data;
+        size_t length = packet->length;
+        const MIDITimeStamp timestamp = packet->timeStamp;
+
+        while (length > 0) {
```

Comment:
Please follow https://docs.libretro.com/development/coding-standards/ https://github.com/libretro/RetroArch/blob/master/CODING-GUIDELINES

Reply:
I'll take care of it later